File: src/shared/lib/sms-parser.ts
/* پارسر پیامک‌های بانکی ایرانی
   تشخیص: نوع (واریز/برداشت)، مبلغ، موجودی، بانک، تاریخ — منطق خالص */

import { toJalali, toGregorian } from "./jalali";
import { toEn } from "./digits";

export type SmsTxType = "income" | "expense";

export interface SmsDate {
  gy: number;
  gm: number;
  gd: number;
  jalali: [number, number, number];
}

export interface ParsedSms {
  rawText: string;
  type: SmsTxType | null;
  bank: string | null;
  amount: number | null;
  balance: number | null;
  date: SmsDate | null;
  category: string | null;
}

const BANKS: { re: RegExp; name: string }[] = [
  { re: /ملی|BMI|bank.?melli/i, name: "بانک ملی" },
  { re: /ملت|BMEL|tejarat.?area/i, name: "بانک ملت" },
  { re: /صادرات|BSIR|bsi/i, name: "بانک صادرات" },
  { re: /تجارت|tejarat/i, name: "بانک تجارت" },
  { re: /پاسارگاد|pasargad/i, name: "بانک پاسارگاد" },
  { re: /پارسیان|parsian/i, name: "بانک پارسیان" },
  { re: /سپه|sepah/i, name: "بانک سپه" },
  { re: /کشاورزی|keshavarzi|bki/i, name: "بانک کشاورزی" },
  { re: /رفاه|refah/i, name: "بانک رفاه" },
  { re: /مسکن|maskan/i, name: "بانک مسکن" },
  { re: /development/i, name: "بانک توسعه صادرات" },
  { re: /سامان|saman/i, name: "بانک سامان" },
  { re: /انصاری|ansari/i, name: "بانک انصاری" },
  { re: /گردشگری|tourism/i, name: "بانک گردشگری" },
  { re: /حکمت|hekmat/i, name: "بانک حکمت" },
  { re: /دی|day/i, name: "بانک دی" },
  { re: /ایران.?زمین|iranzamin/i, name: "بانک ایران‌زمین" },
  { re: /قرض.?الحسنه|qarz|mehr/i, name: "بانک قرض‌الحسنه" },
  { re: /شهر|shahr/i, name: "بانک شهر" },
  { re: /آینده|ayandeh/i, name: "بانک آینده" },
  { re: /سرمایه|sarmayeh/i, name: "بانک سرمایه" },
  { re: /کارآفرین|karafarin/i, name: "بانک کارآفرین" },
  { re: /سینا|sina/i, name: "بانک سینا" },
  { re: /مهر|mehr/i, name: "بانک مهر ایران" },
  { re: /خاورمیانه|khavarmianeh/i, name: "بانک خاورمیانه" },
];

const INCOME_WORDS = [
  "واریز",
  "واریز شد",
  "واریز به",
  "رسید",
  "رسیدید",
  "دریافت",
  "کارمزد واریز",
  "انتقال به حساب شما",
  "به حساب شما واریز",
  "deposit",
  "credited",
];

const EXPENSE_WORDS = [
  "برداشت",
  "برداشت شد",
  "پرداخت",
  "خرید",
  "انتقال",
  "پرداخت اینترنتی",
  "خرید اینترنتی",
  "کارت به کارت",
  "برداشت از",
  "پرداخت قبض",
  "شارژ",
  "withdraw",
  "purchase",
  "payment",
];

/* نرمال‌سازی: ارقام فارسی/عربی + حروف عربی → فارسی + حذف نویسه‌های نامرئی */
export function normalizeSms(s: string): string {
  return toEn(String(s))
    .replace(/ي/g, "ی")
    .replace(/ك/g, "ک")
    .replace(/[\u200c\u200f\u200e]/g, "");
}

/* همه اعداد بزرگ متن (>= 1000) */
export function extractAmounts(text: string): { raw: string; value: number }[] {
  const t = toEn(text);
  const matches = [...t.matchAll(/(\d{1,3}(?:,\d{3})+|\d{4,})(?:\.\d+)?/g)];
  return matches
    .map((m) => ({ raw: m[1], value: +m[1].replace(/,/g, "") }))
    .filter((m) => m.value >= 1000);
}

export function detectType(text: string): SmsTxType | null {
  const t = text.toLowerCase();
  let income = 0,
    expense = 0;
  for (const w of INCOME_WORDS) if (t.includes(w)) income++;
  for (const w of EXPENSE_WORDS) if (t.includes(w)) expense++;
  if (income > expense) return "income";
  if (expense > income) return "expense";
  return null;
}

export function detectBank(text: string): string | null {
  for (const b of BANKS) {
    if (b.re.test(text)) return b.name;
  }
  return null;
}

export function extractDate(text: string): SmsDate | null {
  const t = toEn(text);

  /* جلالی: ۱۴۰۴/۰۶/۱۵ */
  let m = t.match(/1[34]\d{2}[\/\-.](\d{1,2})[\/\-.](\d{1,2})/);
  if (m) {
    const jy = +m[0].slice(0, 4);
    const jm = +m[1],
      jd = +m[2];
    if (jm >= 1 && jm <= 12 && jd >= 1 && jd <= 31) {
      const [gy, gm, gd] = toGregorian(jy, jm, jd);
      return { gy, gm, gd, jalali: [jy, jm, jd] };
    }
  }

  /* میلادی: 2025-09-06 */
  m = t.match(/(20\d{2})[\/\-.](\d{1,2})[\/\-.](\d{1,2})/);
  if (m) {
    const gy = +m[1],
      gm = +m[2],
      gd = +m[3];
    return { gy, gm, gd, jalali: toJalali(gy, gm, gd) };
  }

  /* میلادی با روز اول: 06/09/2025 */
  m = t.match(/(\d{1,2})[\/\-.](\d{1,2})[\/\-.](20\d{2})/);
  if (m) {
    const gy = +m[3],
      gm = +m[2],
      gd = +m[1];
    return { gy, gm, gd, jalali: toJalali(gy, gm, gd) };
  }

  return null;
}

export function extractBalance(text: string): number | null {
  const t = toEn(text);
  let m = t.match(/موجودی[^0-9]{0,15}(\d{1,3}(?:,\d{3})+|\d{4,})/);
  if (m) return +m[1].replace(/,/g, "");
  m = t.match(/مانده[^0-9]{0,15}(\d{1,3}(?:,\d{3})+|\d{4,})/);
  if (m) return +m[1].replace(/,/g, "");
  return null;
}

/* ── حدس دسته‌بندی از متن پیامک ── */
export function guessCategory(text: string): string | null {
  const t = text.toLowerCase();
  if (/قبض|برق|آب|گاز|تلفن|شارژ ماهانه/.test(t)) return "bills";
  if (/خرید|پرداخت اینترنتی|درگاه/.test(t)) return "shopping";
  if (/کارت به کارت|انتقال/.test(t)) return "other-e";
  if (/آموزش|شهریه/.test(t)) return "edu";
  if (/دارو|دکتر|بیمارستان|پزشک/.test(t)) return "health";
  if (/حقوق|salary/.test(t)) return "salary";
  return null;
}

/* ═══════════ پارس کامل ═══════════ */
export function parseSms(rawText: string): ParsedSms | null {
  const text = normalizeSms(rawText || "");
  if (!text) return null;

  const type = detectType(text);
  const bank = detectBank(text);
  const date = extractDate(text);
  const balance = extractBalance(text);
  const amounts = extractAmounts(text);

  /* حذف مبلغ موجودی از لیست مبالغ */
  const txAmounts = amounts.filter((a) => a.value !== balance);
  const amount = txAmounts.length ? txAmounts[0].value : (amounts[0]?.value ?? null);

  return {
    rawText: text,
    type,
    bank,
    amount,
    balance,
    date,
    category: guessCategory(text),
  };
}

/* تفکیک چند پیامک با خط خالی */
export function splitSmsBlocks(raw: string): string[] {
  return raw
    .split(/\n\s*\n/)
    .map((b) => b.trim())
    .filter(Boolean);
}

File: backend/app/services/sms_parser.py
"""پارسرِ پیامکِ بانکیِ ایرانی — پورتِ منطقِ خالصِ src/shared/lib/sms-parser.ts.

چرا سمت سرور؟ تا پیش از این، پارسِ پیامک «فقط» در کلاینت (TS) انجام می‌شد و مسیرهای
مختلفِ ورودی (فورواردرِ خودکار که فقط متنِ خام را ذخیره می‌کرد، در برابر مسیرِ دستی که
سمت کلاینت پارس می‌شد) خروجیِ ناهم‌سان می‌دادند. با انتقالِ پارس به سرور، هر دو مسیرِ
ورود (خواندنِ نیتیوِ اپِ اندروید + پلِ فورواردر) دقیقاً یک‌جور تراکنش می‌سازند و وب هم
همان نتیجه را می‌بیند.

بدونِ هیچ وابستگیِ بیرونی؛ فقط کتابخانهٔ استاندارد + app.services.jalali.
منطق مو‌به‌مو با نسخهٔ TS یکی است تا در دورهٔ گذار (که هر دو زنده‌اند) اختلافی نباشد.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import date

from app.services.jalali import to_gregorian, to_jalali

# ── ارقام فارسی/عربی → انگلیسی (معادلِ digits.ts:toEn) ───────────
_FA_DIGITS = "۰۱۲۳۴۵۶۷۸۹"
_AR_DIGITS = "٠١٢٣٤٥٦٧٨٩"
_DIGIT_MAP: dict[int, str] = {ord(c): str(i) for i, c in enumerate(_FA_DIGITS)}
_DIGIT_MAP.update({ord(c): str(i) for i, c in enumerate(_AR_DIGITS)})


def to_en(value: str) -> str:
    """ارقام فارسی/عربی → انگلیسی. سایرِ نویسه‌ها دست‌نخورده."""
    return str(value).translate(_DIGIT_MAP)


# ── بانک‌ها (ترتیب مهم است؛ اولین تطبیق برنده) ──────────────────
_BANKS: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"ملی|BMI|bank.?melli", re.I), "بانک ملی"),
    (re.compile(r"ملت|BMEL|tejarat.?area", re.I), "بانک ملت"),
    (re.compile(r"صادرات|BSIR|bsi", re.I), "بانک صادرات"),
    (re.compile(r"تجارت|tejarat", re.I), "بانک تجارت"),
    (re.compile(r"پاسارگاد|pasargad", re.I), "بانک پاسارگاد"),
    (re.compile(r"پارسیان|parsian", re.I), "بانک پارسیان"),
    (re.compile(r"سپه|sepah", re.I), "بانک سپه"),
    (re.compile(r"کشاورزی|keshavarzi|bki", re.I), "بانک کشاورزی"),
    (re.compile(r"رفاه|refah", re.I), "بانک رفاه"),
    (re.compile(r"مسکن|maskan", re.I), "بانک مسکن"),
    (re.compile(r"development", re.I), "بانک توسعه صادرات"),
    (re.compile(r"سامان|saman", re.I), "بانک سامان"),
    (re.compile(r"انصاری|ansari", re.I), "بانک انصاری"),
    (re.compile(r"گردشگری|tourism", re.I), "بانک گردشگری"),
    (re.compile(r"حکمت|hekmat", re.I), "بانک حکمت"),
    (re.compile(r"دی|day", re.I), "بانک دی"),
    (re.compile(r"ایران.?زمین|iranzamin", re.I), "بانک ایران‌زمین"),
    (re.compile(r"قرض.?الحسنه|qarz|mehr", re.I), "بانک قرض‌الحسنه"),
    (re.compile(r"شهر|shahr", re.I), "بانک شهر"),
    (re.compile(r"آینده|ayandeh", re.I), "بانک آینده"),
    (re.compile(r"سرمایه|sarmayeh", re.I), "بانک سرمایه"),
    (re.compile(r"کارآفرین|karafarin", re.I), "بانک کارآفرین"),
    (re.compile(r"سینا|sina", re.I), "بانک سینا"),
    (re.compile(r"مهر|mehr", re.I), "بانک مهر ایران"),
    (re.compile(r"خاورمیانه|khavarmianeh", re.I), "بانک خاورمیانه"),
]

_INCOME_WORDS = [
    "واریز",
    "واریز شد",
    "واریز به",
    "رسید",
    "رسیدید",
    "دریافت",
    "کارمزد واریز",
    "انتقال به حساب شما",
    "به حساب شما واریز",
    "deposit",
    "credited",
]

_EXPENSE_WORDS = [
    "برداشت",
    "برداشت شد",
    "پرداخت",
    "خرید",
    "انتقال",
    "پرداخت اینترنتی",
    "خرید اینترنتی",
    "کارت به کارت",
    "برداشت از",
    "پرداخت قبض",
    "شارژ",
    "withdraw",
    "purchase",
    "payment",
]

# ── الگوهای عدد/تاریخ/کلیدواژه ──────────────────────────────────
_NUM = r"(\d{1,3}(?:,\d{3})+|\d{4,})"
_AMOUNT_RE = re.compile(_NUM + r"(?:\.\d+)?")
_AMOUNT_KW_RE = re.compile(r"مبلغ[^0-9]{0,10}" + _NUM)
_BAL_MOJUDI_RE = re.compile(r"موجودی[^0-9]{0,15}" + _NUM)
_BAL_MANDE_RE = re.compile(r"مانده[^0-9]{0,15}" + _NUM)
_JALALI_RE = re.compile(r"1[34]\d{2}[/\-.](\d{1,2})[/\-.](\d{1,2})")
_GREG_RE = re.compile(r"(20\d{2})[/\-.](\d{1,2})[/\-.](\d{1,2})")
_GREG_DAYFIRST_RE = re.compile(r"(\d{1,2})[/\-.](\d{1,2})[/\-.](20\d{2})")
_INVISIBLE_RE = re.compile(r"[‌‏‎]")

_CAT_RULES: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"قبض|برق|آب|گاز|تلفن|شارژ ماهانه"), "bills"),
    (re.compile(r"خرید|پرداخت اینترنتی|درگاه"), "shopping"),
    (re.compile(r"کارت به کارت|انتقال"), "other-e"),
    (re.compile(r"آموزش|شهریه"), "edu"),
    (re.compile(r"دارو|دکتر|بیمارستان|پزشک"), "health"),
    (re.compile(r"حقوق|salary"), "salary"),
]


@dataclass(frozen=True)
class SmsDate:
    """تاریخِ استخراج‌شده — میلادی (gy/gm/gd) به‌همراهِ معادلِ جلالی."""

    gy: int
    gm: int
    gd: int
    jalali: tuple[int, int, int]

    def to_date(self) -> date:
        """datetime.date میلادی؛ در صورتِ اجزای نامعتبر ValueError می‌دهد."""
        return date(self.gy, self.gm, self.gd)


@dataclass(frozen=True)
class ParsedSms:
    """نتیجهٔ خالصِ پارس. category ستونِ دیتابیس نیست (هنگام تبدیلِ پیامک به تراکنش
    در کلاینت به‌کار می‌رود) و صرفاً برای اطلاع برگردانده می‌شود."""

    normalized: str
    type: str | None
    bank: str | None
    amount: int | None
    balance: int | None
    date: date | None
    category: str | None


def normalize_sms(s: str) -> str:
    """ارقام فارسی/عربی → انگلیسی + ي→ی، ك→ک + حذفِ نویسه‌های نامرئی."""
    return _INVISIBLE_RE.sub("", to_en(str(s)).replace("ي", "ی").replace("ك", "ک"))


def extract_amounts(text: str) -> list[int]:
    """همهٔ اعدادِ «بزرگ» (>= ۱۰۰۰) به ترتیبِ ظهور."""
    t = to_en(text)
    out: list[int] = []
    for m in _AMOUNT_RE.finditer(t):
        v = int(m.group(1).replace(",", ""))
        if v >= 1000:
            out.append(v)
    return out


def detect_type(text: str) -> str | None:
    """رأی‌گیریِ کلیدواژه‌ای: واریز/درآمد در برابر برداشت/هزینه."""
    t = text.lower()
    income = sum(1 for w in _INCOME_WORDS if w in t)
    expense = sum(1 for w in _EXPENSE_WORDS if w in t)
    if income > expense:
        return "income"
    if expense > income:
        return "expense"
    return None


def detect_bank(text: str) -> str | None:
    for pat, name in _BANKS:
        if pat.search(text):
            return name
    return None


def extract_date(text: str) -> SmsDate | None:
    """جلالی (۱۴xx/xx/xx) → میلادی؛ وگرنه میلادیِ 20xx (سال-اول یا روز-اول)."""
    t = to_en(text)
    m = _JALALI_RE.search(t)
    if m:
        jy = int(m.group(0)[:4])
        jm, jd = int(m.group(1)), int(m.group(2))
        if 1 <= jm <= 12 and 1 <= jd <= 31:
            gy, gm, gd = to_gregorian(jy, jm, jd)
            return SmsDate(gy, gm, gd, (jy, jm, jd))
        # جلالیِ خارج از بازه: به شاخهٔ میلادی می‌افتد (مثلِ نسخهٔ TS)
    m = _GREG_RE.search(t)
    if m:
        gy, gm, gd = int(m.group(1)), int(m.group(2)), int(m.group(3))
        return SmsDate(gy, gm, gd, to_jalali(gy, gm, gd))
    m = _GREG_DAYFIRST_RE.search(t)
    if m:
        gy, gm, gd = int(m.group(3)), int(m.group(2)), int(m.group(1))
        return SmsDate(gy, gm, gd, to_jalali(gy, gm, gd))
    return None


def extract_balance(text: str) -> int | None:
    """موجودی/مانده پس از کلیدواژه (تا ۱۵ نویسهٔ غیرعددی فاصله)."""
    t = to_en(text)
    m = _BAL_MOJUDI_RE.search(t)
    if m:
        return int(m.group(1).replace(",", ""))
    m = _BAL_MANDE_RE.search(t)
    if m:
        return int(m.group(1).replace(",", ""))
    return None


def guess_category(text: str) -> str | None:
    t = text.lower()
    for pat, cat in _CAT_RULES:
        if pat.search(t):
            return cat
    return None


def parse_sms(raw_text: str) -> ParsedSms | None:
    """پارسِ کاملِ یک پیامک. None اگر متن پس از نرمال‌سازی خالی باشد.

    مبلغِ تراکنش به ترتیبِ اولویت:
      ۱) عددِ پس از کلیدواژهٔ «مبلغ» (اگر با موجودی برابر نباشد)،
      ۲) اولین عددِ بزرگِ ≠ موجودی،
      ۳) اولین عددِ بزرگ.
    """
    text = normalize_sms(raw_text or "")
    if not text:
        return None

    stype = detect_type(text)
    bank = detect_bank(text)
    sms_date = extract_date(text)
    balance = extract_balance(text)
    amounts = extract_amounts(text)

    amount: int | None = None
    kw = _AMOUNT_KW_RE.search(text)
    if kw:
        v = int(kw.group(1).replace(",", ""))
        if v != balance:
            amount = v
    if amount is None:
        tx_amounts = [a for a in amounts if a != balance]
        amount = tx_amounts[0] if tx_amounts else (amounts[0] if amounts else None)

    parsed_date: date | None = None
    if sms_date is not None:
        try:
            parsed_date = sms_date.to_date()
        except ValueError:
            parsed_date = None  # اجزای تاریخِ نامعتبر → بدونِ تاریخ (به‌جای خطا)

    return ParsedSms(
        normalized=text,
        type=stype,
        bank=bank,
        amount=amount,
        balance=balance,
        date=parsed_date,
        category=guess_category(text),
    )


def split_sms_blocks(raw: str) -> list[str]:
    """تفکیکِ چند پیامک با خطِ خالی (معادلِ splitSmsBlocks)."""
    return [b.strip() for b in re.split(r"\n\s*\n", raw) if b.strip()]
